qml/pages/MessageListView.qml

```diff
@@ -14,7 +14,41 @@ SilicaListView {
     property bool hasMoreMessages: false
     property bool loading: false
     property bool canLoadMore: hasMoreMessages && !loading
-    property string latestRead: ""
+    // TODO update this based on scroll
+    property string furthestRead: ""
+    property string currentLastRead: channel.lastRead
+
+    // C++-ify this
+    function timestampToIndex(timestamp) {
+        // TODO: better than linear search for message to be updated
+        for (var i = 0; i < messageListModel.count; i++) {
+            if (messageListModel.get(i).timestamp === timestamp) {
+                return i
+            }
+        }
+        return -1
+    }
+
+    function setLastRead(timestamp, index) {
+        console.log("Setting lastRead to", timestamp)
+        if (typeof index === "number") {
+            // If we're passing an index it means that the _previous_ message needs to be marked.
+            if (index > 0) {
+                var firstUnreadMessage = messageListModel.get(index - 1);
+                timestamp = firstUnreadMessage.timestamp;
+            } else {
+                timestamp = "0000000000.000000"
+            }
+        }
```

Comment:
The `index` only seems to be used when calling this function from the callback of "Mark unread" message action (which makes sense - you want to mark only the previous message as read in that case). Would it make sense to move this code into the `onMarkUnread` slot, so that the function signature becomes simpler and the behavior of the function more predictable?

Reply:
Good idea!